docs/extending/extending_client_side.md

```diff
@@ -0,0 +1,528 @@
+            this.updateCount();
+        }
+
+        setupOutput() {
+            if (this.output) return;
+            const template = document.createElement('template');
+            template.innerHTML = `<output name='word-count' for='${this.element.id}' style='float: right;'></output>`;
+            const output = template.content.firstChild;
+            this.element.insertAdjacentElement('beforebegin', output);
+            this.output = output;
+        }
+
+        updateCount(event) {
+            const value = event ? event.target.value : this.element.value;
+            const words = (value || '').split(' ');
+            this.output.textContent = `${words.length} / ${this.maxValue} words`;
+        }
+
+        disconnect() {
+            this.element && this.element.remove();
+        }
+    }
+    Stimulus.register('word-count', WordCountController);
+})(window.Stimulus);
+```
+
+This lets the data attribute values determine the 'configuration' of this controller and the data attribute actions to determine the 'triggers' for the updates to the output element.
+
+```python
+# models.py
+from django import forms
+
+from wagtail.admin.panels import FieldPanel
+from wagtail.models import Page
+
+
+class BlogPage(Page):
+    # ...
+    content_panels = Page.content_panels + [
+        FieldPanel('subtitle', classname="full"),
+        FieldPanel(
+            'introduction',
+            classname="full",
+            widget=forms.TextInput(
+                attrs={
+                    'data-controller': 'word-count',
+                    # allow the max number to be determined with attributes
+                    # note we can use Python values here, Django will handle the string conversion (including escaping if applicable)
+                    'data-word-count-max-value': 5,
+                    # decide when you want the count to update with data-action
+                    # (e.g. 'blur->word-count#updateCount' will only update when field loses focus)
+                    'data-action': 'word-count#updateCount paste->word-count#updateCount',
+                }
+            )
+        ),
+    #...
+```
+
+This next code snippet shows a slightly more advanced version of the `insert_editor_js` hook usage which is set up to easily append additional scripts for future controllers.
+
+```python
+# wagtail_hooks.py
+from django.utils.html import format_html_join
+from django.templatetags.static import static
+
+from wagtail import hooks
+
+
+@hooks.register('insert_editor_js')
+def editor_js():
+    # add more controller code as needed
+    js_files = ['js/word-count-controller.js',]
+    return format_html_join('\n', '<script src="{0}"></script>',
+        ((static(filename),) for filename in js_files)
+    )
+```
+
+You should be able to see that on your Blog Page, all introduction fields will now have a small `output` element showing the count and max words being used.
+
+#### A word count controller example with a build system
```

Comment:
This strikes me as way too much information compared to what we definitely need to convey here:

- Use Wagtail’s Stimulus package with our global variable(s) if possible, or your own copy of `@hotwired/stimulus` if needed.
- Make sure to use a build target higher than ES5